### DISCUSSION TITLE
stateful=links is not working b setting workers other than one

### DISCUSSION THREAD

**Discussion by hegebharat** [schemathesis/schemathesis#1132] [General]
Hello ,

I am using schemathesis CLI.
In-order to speed up the schemathesis execution i am setting workers to 4 also --stateful=links is set.

After setting workers to 4(anything other than 1)  schemathesis is not generating the scenarios. So I have few questions.

1) is stateful testing is disabled if we set workers other-than 1? If not what could be the reason in my case.
2) In callbacks.py  os.cpu_count is used if we auto option. Here why its checking the number of cpu instead of number of threads?

Thanks you

**Comment by Stranger6667** [comment]
Hi!

> is stateful testing is disabled if we set workers other-than 1? If not what could be the reason in my case.

No, it is not disabled. There is [a couple](https://github.com/schemathesis/schemathesis/blob/a2b0de3e572e1160b84cb4c4265bd119d8a43379/test/cli/test_commands.py#L1601) of tests for that behavior. Could you, please describe what happens in your case? I.e. what behavior you considered as Schemathesis not generating scenarios?

> In callbacks.py os.cpu_count is used if we auto option. Here why its checking the number of cpu instead of number of threads?

`os.cpu_count` is only used as a fallback for systems without support for `os.sched_getaffinity`. I.e. on my machine that has [6 physical cores](https://ark.intel.com/content/www/us/en/ark/products/126684/intel-core-i7-8700k-processor-12m-cache-up-to-4-70-ghz.html), the `get_workers_count` returns 12 which corresponds to the number of threads mentioned in the CPU spec. Both, `os.cpu_count` and `os.sched_getaffinity` are actually returning this number of logical CPUs (at least on my machine). However, I believe that it might be platform-dependent - maybe on some systems it actually returns the number of physical cores (I am not sure about that) and I missed that.

**Reply by hegebharat** [reply]
Hi, 
![image](https://user-images.githubusercontent.com/27925587/114844520-7f697500-9df8-11eb-87b1-051ecc6ba9c0.png)

after increasing the number of workers to 3, I cannot the endpoint details in the output console(as you can see dotted line). ALso total number of request in decreased(I repeated 3 time the experiment) compared to experimenting with workers to 1 .  Therefor eI am not sure whether the scenarios(stateful links) is executed or not. I also checked the output log file to see anyway if I can track the scenarios.

So
1)  how do I verify that scenarios are executed  after setting more than one workers. 
2) IF for example consider the scenario(api1->ap2->api3). api1 return 200 and response of api1 is used by api2 but lets say api2 is return 404. How can I keep track of at which api call this scenario is failed. 


Thanks a lot.

**Reply by Stranger6667** [reply]
Hi!

Sorry for the delay.

> after increasing the number of workers to 3, I cannot the endpoint details in the output console(as you can see dotted line).

Yep, this output style is specific to multi-worker config - unfortunately, I didn't come up with a different output style that will reflect stateful test output when multiple workers are set.

> ALso total number of request in decreased(I repeated 3 time the experiment) compared to experimenting with workers to 1 .

Hmm, it is strange, could you share your schema and command-line options used?
With the schema from the Schemathesis test suite, I see the same number of calls executed (the number of performed checks corresponds to the number of requests there).

The following routine executed multiple times with the latest Schemathesis version reports the same number of requests for single-worker and multi-worker config.

Single worker:
1. Go to the Schemathesis repo root
2. Remove Hypothesis db - `rm -rf .hypothesis`
3. Start the test server - `./test_server.sh 8081  --spec=openapi3 --operations=create_user,get_user,update_user`
4. Run Schemathesis - `schemathesis run http://127.0.0.1:8081/schema.yaml --stateful=links --hypothesis-seed=42`
5. Observe `not_a_server_error                    600 / 600 passed          PASSED` in the output

Multiple workers are the same except `-w 3` is added to the command from the 4th step.
The number of executed requests depends also on the app's internal state - so it should be cleaned up between runs (e.g. to avoid duplicate objects created in the app's DB, etc).

Those stateful sequences should be executed, but the order in which it is saved to e.g. VCR cassette is not preserved - individual requests/responses are saved in the order they came to the cassette recorder.

> how do I verify that scenarios are executed after setting more than one workers.

Unfortunately, it is not directly possible to track those sequences :( Internal events contain the `thread_id` field which could use to filter requests/responses coming from the same worker. The order of requests that come from the same worker is preserved though. Additionally, each case stores its `source` which could also be used to track sequences. Then I can propose the following changes to Schemathesis output logs to simplify the tracking of those sequences:
- Store `thread_id` for each request/response pair;
- Store `AfterExecution.correlation_id` for each pair;
- Introduce optional `source.correlation_id` that will contain the `AfterExecution.correlation_id` value of the "source" response

At this moment it seems to be sufficient to track all sequences in logs. What do you think?
But, of course, it will be also nice to have a nice stdout style for such a case - but I ran out of ideas for that. Also, it worth mentioning that in the future stateful testing used in CLI will be replaced with the one based on Hypothesis state machines - at the moment I don't know how to have not overly verbose and ergonomic output :( 

Suggestions and ideas are welcome :)

> IF for example consider the scenario(api1->ap2->api3). api1 return 200 and response of api1 is used by api2 but lets say api2 is return 404. How can I keep track of at which api call this scenario is failed.

I think that the situation is the same for such a use case :( But those potential changes to logs should allow you to track those by analyzing the logs. Other thoughts on the output style apply here as well.

**Reply by hegebharat** [reply]
Sorry for the late reply. Thanks a lot for the detailed explanation.  One idea may be during feedback module simply add the parent info(main case object from which link is generated) to track the feedback links also actually which data (id, name) is parsed between the call. This might useful to the developer see why something went wrong when we talk about multiple calls.
